core/src/wsv.rs

```diff
@@ -411,59 +397,32 @@ impl WorldStateView {
         self.asset(id).map_err(Into::into)
     }
 
-    /// Update metrics; run when block commits.
-    fn block_commit_metrics_update_callback(&self) {
-        let last_block_txs_accepted = self
-            .blocks
-            .iter()
-            .last()
-            .map(|block| block.as_v1().transactions.len() as u64)
-            .unwrap_or_default();
-        let last_block_txs_rejected = self
-            .blocks
-            .iter()
-            .last()
-            .map(|block| block.as_v1().rejected_transactions.len() as u64)
-            .unwrap_or_default();
-        self.metrics
-            .txs
-            .with_label_values(&["accepted"])
-            .inc_by(last_block_txs_accepted);
-        self.metrics
-            .txs
-            .with_label_values(&["rejected"])
-            .inc_by(last_block_txs_rejected);
-        self.metrics
-            .txs
-            .with_label_values(&["total"])
-            .inc_by(last_block_txs_accepted + last_block_txs_rejected);
-        self.metrics.block_height.inc();
-    }
-
-    // TODO: There could be just this one method `blocks` instead of
-    // `blocks_from_height` and `blocks_after_height`. Also, this
-    // method would return references instead of cloning blockchain
-    // but comes with the risk of deadlock if consumer of the iterator
-    // stores references to blocks
-    /// Returns iterator over blockchain blocks
-    ///
-    /// **Locking behaviour**: Holding references to blocks stored in the blockchain can induce
-    /// deadlock. This limitation is imposed by the fact that blockchain is backed by [`dashmap::DashMap`]
-    #[inline]
-    pub fn blocks(&self) -> crate::block::ChainIterator {
-        self.blocks.iter()
+    /// Load all blocks in the block chain from disc and clone them for use.
+    pub fn all_blocks_by_value(&self) -> Vec<VersionedCommittedBlock>
+    {
+        let mut blocks_array = Vec::new();
+        let block_count = self.block_hashes.borrow().len() as u64;
+        for i in 0..block_count {
+            blocks_array.push(VersionedCommittedBlock::clone(&self.kura.get_block_by_height(i + 1).expect("Failed to load block.")));
+        }
+        blocks_array
     }
 
     /// Return a vector of blockchain blocks after the block with the given `hash`
-    pub fn blocks_after_hash(
+    pub fn block_hashes_after_hash(
         &self,
         hash: HashOf<VersionedCommittedBlock>,
-    ) -> Vec<VersionedCommittedBlock> {
-        self.blocks
-            .iter()
-            .skip_while(move |block_entry| block_entry.value().header().previous_block_hash != hash)
-            .map(|block_entry| block_entry.value().clone())
-            .collect()
+    ) -> Vec<HashOf<VersionedCommittedBlock>> {
+        if hash == Hash::zeroed().typed() {
```

Comment:
Most hashes are non-zeroed. We also store a lot of them. I'd consider declaring a niche, and turning the few places where hashes can also be zeroed into an `Option<Hash>`.